# MONGODB

I. Installing Pymongo and installing required libraries

In [ ]:
# Install PyMongo
!pip install pymongo

# Import libraries
from pymongo import MongoClient
import pandas as pd
import json

print("Libraries loaded successfully")

Libraries loaded successfully


II. Loading Cleaned Data from GitHub

In [ ]:
# Clone GitHub repo
!git clone https://github.com/nomankaleemromane/NorthStar-DBA-Coursework.git

import pandas as pd

# Load all cleaned CSV files
customers_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/customers_cleaned.csv")
orders_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/orders_cleaned.csv")
deliveries_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/deliveries_cleaned.csv")
drivers_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/drivers_cleaned.csv")
vehicles_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/vehicles_cleaned.csv")
hubs_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/hubs_cleaned.csv")
complaints_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/complaints_cleaned.csv")
incidents_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/incidents_cleaned.csv")
app_events_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/app_events_cleaned.csv")
data_dictionary_df = pd.read_csv("NorthStar-DBA-Coursework/cleaned_dataset/data_dictionary_cleaned.csv")

print("All cleaned files loaded successfully")

fatal: destination path 'NorthStar-DBA-Coursework' already exists and is not an empty directory.
All cleaned files loaded successfully


III. Connecting to MongoDB Atlas

In [ ]:
# Connect to MongoDB Atlas
connection_string = "mongodb+srv://32146940_db_user:32146940_db_user@dba.rv4llxi.mongodb.net/?appName=DBA"
client = MongoClient(connection_string)

# Create database
db = client["NorthStarDB"]

# Verify connection
print("Connected to MongoDB Atlas")
print("Databases:", client.list_database_names())

Connected to MongoDB Atlas
Databases: ['NorthStarDB', 'admin', 'local']


IV. Inserting Data into Collections

1. customers collection


In [ ]:
customers_col = db["customers"]

customer_docs = []
for _, cust in customers_df.iterrows():
    cust_complaints = complaints_df[complaints_df['customer_id'] == cust['customer_id']].to_dict('records')
    cust_events = app_events_df[app_events_df['customer_id'] == cust['customer_id']].to_dict('records')
    cust_orders = orders_df[orders_df['customer_id'] == cust['customer_id']].to_dict('records')

    doc = {
        "_id": cust['customer_id'],
        "age": cust['age'],
        "home_zone": cust['home_zone'],
        "customer_type": cust['customer_type'],
        "loyalty_score": cust['loyalty_score'],
        "contact": {
            "preferred_channel": cust['preferred_channel'],
            "account_status": cust['account_status']
        },
        "orders": cust_orders,
        "complaints": cust_complaints,
        "app_events": cust_events
    }
    customer_docs.append(doc)

customers_col.insert_many(customer_docs)
print(f"Inserted {len(customer_docs)} customer documents")

Inserted 650 customer documents


2. drivers collection


In [ ]:
drivers_col = db["drivers"]

driver_docs = []
for _, drv in drivers_df.iterrows():
    driver_deliveries = deliveries_df[deliveries_df['driver_id'] == drv['driver_id']]
    driver_incidents = []
    for _, delv in driver_deliveries.iterrows():
        inc = incidents_df[incidents_df['delivery_id'] == delv['delivery_id']]
        driver_incidents.extend(inc.to_dict('records'))

    doc = {
        "_id": drv['driver_id'],
        "base_zone": drv['base_zone'],
        "employment_type": drv['employment_type'],
        "years_experience": drv['years_experience'],
        "training_score": drv['training_score'],
        "rating": drv['driver_rating'],
        "shift_preference": drv['shift_preference'],
        "active": bool(drv['active_flag']),
        "deliveries": driver_deliveries.to_dict('records'),
        "incidents": driver_incidents
    }
    driver_docs.append(doc)

drivers_col.insert_many(driver_docs)
print(f"Inserted {len(driver_docs)} driver documents")

Inserted 170 driver documents


3. vehicles section


In [ ]:
vehicles_col = db["vehicles"]

vehicle_docs = []
for _, veh in vehicles_df.iterrows():
    vehicle_deliveries = deliveries_df[deliveries_df['vehicle_id'] == veh['vehicle_id']]
    vehicle_incidents = []
    for _, delv in vehicle_deliveries.iterrows():
        inc = incidents_df[incidents_df['delivery_id'] == delv['delivery_id']]
        vehicle_incidents.extend(inc.to_dict('records'))

    doc = {
        "_id": veh['vehicle_id'],
        "vehicle_type": veh['vehicle_type'],
        "assigned_zone": veh['assigned_zone'],
        "commission_date": veh['commission_date'],
        "battery_health_pct": veh['battery_health_pct'],
        "odometer_km": veh['odometer_km'],
        "maintenance_status": veh['maintenance_status'],
        "maintenance_log": [],  # Populated as maintenance records become available
        "sensor_alerts": [],     # Populated from IoT sensor data
        "incidents": vehicle_incidents
    }
    vehicle_docs.append(doc)

vehicles_col.insert_many(vehicle_docs)
print(f"Inserted {len(vehicle_docs)} vehicle documents")

Inserted 120 vehicle documents


4. deliveries collection

In [ ]:
deliveries_col = db["deliveries"]

delivery_docs = []
for _, delv in deliveries_df.iterrows():
    delivery_incidents = incidents_df[incidents_df['delivery_id'] == delv['delivery_id']]

    doc = {
        "_id": delv['delivery_id'],
        "order_id": delv['order_id'],
        "driver_id": delv['driver_id'],
        "vehicle_id": delv['vehicle_id'],
        "hub_id": delv['hub_id'],
        "dispatch_time": delv['dispatch_time'],
        "completed_at": delv['delivery_completed_at'],
        "status": delv['delivery_status'],
        "route": {
            "distance_km": delv['route_distance_km'],
            "manual_overrides": delv['manual_route_override_count'],
            "override_details": []  # Populated from driver app route change logs
        },
        "cost": {
            "fuel_or_charge": delv['fuel_or_charge_cost']
        },
        "customer_rating": delv['customer_rating_post_delivery'],
        "proof_of_completion": {
            "missing": bool(delv['proof_of_completion_missing']),
            "reason": None
        },
        "exceptions": delivery_incidents.to_dict('records')
    }
    delivery_docs.append(doc)

deliveries_col.insert_many(delivery_docs)
print(f"Inserted {len(delivery_docs)} delivery documents")

Inserted 950 delivery documents


app_events

In [ ]:
app_events_col = db["app_events"]
app_event_docs = app_events_df.to_dict('records')
app_events_col.insert_many(app_event_docs)
print(f"Inserted {len(app_event_docs)} app event documents")

Inserted 640 app event documents


V. CRUD Operations

1. FIND — Retrieve Documents

    a. Find a customer by ID:

In [ ]:
# Retrieve a single customer document
customer = customers_col.find_one({"_id": "C0464"})
print(json.dumps(customer, indent=2, default=str))

{
  "_id": "C0464",
  "age": 71,
  "home_zone": "Airport",
  "customer_type": "Enterprise",
  "loyalty_score": 28.8,
  "contact": {
    "preferred_channel": "Phone",
    "account_status": "Dormant"
  },
  "orders": [
    {
      "order_id": "O00168",
      "customer_id": "C0464",
      "service_type": "Parcel",
      "order_created_at": "2025-06-11 03:29:00",
      "promised_window_hours": 6,
      "pickup_zone": "Central",
      "dropoff_zone": "Riverside",
      "priority_level": "High",
      "order_value": 115.35,
      "booking_channel": "Web",
      "special_handling_flag": 0
    },
    {
      "order_id": "O00814",
      "customer_id": "C0464",
      "service_type": "Passenger",
      "order_created_at": "2025-03-26 02:36:00",
      "promised_window_hours": 12,
      "pickup_zone": "East",
      "dropoff_zone": "Riverside",
      "priority_level": "Medium",
      "order_value": 27.9,
      "booking_channel": "Phone",
      "special_handling_flag": 0
    }
  ],
  "complaints": [


    b. Find all failed deliveries:



In [ ]:
# Find all failed deliveries
failed = deliveries_col.find({"status": "Failed"})
print(f"Failed deliveries: {deliveries_col.count_documents({'status': 'Failed'})}")

Failed deliveries: 132


    c. Find complaints by severity:



In [ ]:
# Find high severity complaints
high_complaints = customers_col.find({"complaints.severity": "High"})
print(f"Customers with high severity complaints: {len(list(high_complaints))}")

Customers with high severity complaints: 69


2. INSERT — Add New Documents

    a. Insert a new customer:

In [ ]:
#adding a new customer
new_customer = {
    "_id": "C0651",
    "age": 34,
    "home_zone": "North",
    "customer_type": "Consumer",
    "loyalty_score": 75.5,
    "contact": {
        "preferred_channel": "App",
        "account_status": "Active"
    },
    "orders": [],
    "complaints": [],
    "app_events": []
}

customers_col.insert_one(new_customer)
print(f"Inserted new customer: {customers_col.find_one({'_id': 'C0651'})['_id']}")

Inserted new customer: C0651


    b. Insert a new delivery with an exception:



In [ ]:
#adding new delivery with exception
new_delivery = {
    "_id": "DL00951",
    "order_id": "O01251",
    "driver_id": "D001",
    "vehicle_id": "V001",
    "hub_id": "H01",
    "dispatch_time": "2026-05-01T08:00:00Z",
    "completed_at": "2026-05-01T18:00:00Z",
    "status": "Delayed",
    "route": {
        "distance_km": 15.5,
        "manual_overrides": 2,
        "override_details": [
            {"timestamp": "2026-05-01T09:30:00Z", "reason": "RoadClosed", "new_path": "A4"}
        ]
    },
    "cost": {"fuel_or_charge": 14.50},
    "customer_rating": None,
    "proof_of_completion": {"missing": False, "reason": None},
    "exceptions": [
        {"type": "LateArrival", "recorded_at": "2026-05-01T18:00:00Z", "description": "Delivery outside promised window"}
    ]
}

deliveries_col.insert_one(new_delivery)
print(f"Inserted new delivery: {deliveries_col.find_one({'_id': 'DL00951'})['_id']}")

Inserted new delivery: DL00951


3. UPDATE - Modify Existing Documents

    a. Update a delivery status:

In [ ]:
deliveries_col.update_one(
    {"_id": "DL00951"},
    {"$set": {"status": "OnTime", "customer_rating": 4.5}}
)
print("Updated delivery status:")
print(json.dumps(deliveries_col.find_one({"_id": "DL00951"}), indent=2, default=str))

Updated delivery status:
{
  "_id": "DL00951",
  "order_id": "O01251",
  "driver_id": "D001",
  "vehicle_id": "V001",
  "hub_id": "H01",
  "dispatch_time": "2026-05-01T08:00:00Z",
  "completed_at": "2026-05-01T18:00:00Z",
  "status": "OnTime",
  "route": {
    "distance_km": 15.5,
    "manual_overrides": 2,
    "override_details": [
      {
        "timestamp": "2026-05-01T09:30:00Z",
        "reason": "RoadClosed",
        "new_path": "A4"
      }
    ]
  },
  "cost": {
    "fuel_or_charge": 14.5
  },
  "customer_rating": 4.5,
  "proof_of_completion": {
    "missing": false,
    "reason": null
  },
  "exceptions": [
    {
      "type": "LateArrival",
      "recorded_at": "2026-05-01T18:00:00Z",
      "description": "Delivery outside promised window"
    }
  ]
}


    c. Add a complaint to an existing customer:

In [ ]:
new_complaint = {
    "complaint_id": "CP0321",
    "type": "Delay",
    "severity": "Medium",
    "channel": "App",
    "status": "Open",
    "messages": [
        {"timestamp": "2026-05-05T10:00:00Z", "sender": "customer", "text": "Order not delivered on time"}
    ],
    "resolution_days": None,
    "compensation": 0
}

customers_col.update_one(
    {"_id": "C0651"},
    {"$push": {"complaints": new_complaint}}
)
print("Added complaint to customer")

Added complaint to customer


4. DELETE - Remove Documents

    a. Remove the test customer:



In [ ]:
customers_col.delete_one({"_id": "C0651"})
print(f"Customer deleted. Remaining: {customers_col.count_documents({})}")

Customer deleted. Remaining: 650


    b. Remove the test delivery:

In [ ]:
deliveries_col.delete_one({"_id": "DL00951"})
print(f"Delivery deleted. Remaining: {deliveries_col.count_documents({})}")

Delivery deleted. Remaining: 950


5. Complex Query  Aggregation

Count complaints by type across all customers:

In [ ]:
pipeline = [
    {"$unwind": "$complaints"},
    {"$group": {"_id": "$complaints.complaint_type", "count": {"$sum": 1}}},
    {"$sort": {"count": -1}}
]

result = list(customers_col.aggregate(pipeline))
for r in result:
    print(f"{r['_id']}: {r['count']}")

Delay: 101
MissedPickup: 64
AppIssue: 53
DriverBehaviour: 51
SupportExperience: 20
Billing: 16
Damage: 15


VI. Query Optimisation — MongoDB Indexing

Performance Before Indexing

In [ ]:
import time

# Aggregation WITHOUT any custom indexes
start = time.time()
pipeline = [
    {"$unwind": "$complaints"},
    {"$group": {"_id": "$complaints.complaint_type", "count": {"$sum": 1}}},
    {"$sort": {"count": -1}}
]
result = list(customers_col.aggregate(pipeline))
end = time.time()
before_time = round((end - start) * 1000, 2)
print(f"Without indexes: {before_time} ms")

Without indexes: 79.48 ms


Creating Indexes

In [ ]:
# Create indexes on frequently queried fields
customers_col.create_index("home_zone")
customers_col.create_index("complaints.complaint_type")
customers_col.create_index("complaints.severity")
deliveries_col.create_index("status")
deliveries_col.create_index("driver_id")
deliveries_col.create_index("hub_id")
deliveries_col.create_index("vehicle_id")

# Verify all indexes
print("Indexes on customers:")
for index in customers_col.list_indexes():
    print(f"  - {index['name']}: {index['key']}")

print("\nIndexes on deliveries:")
for index in deliveries_col.list_indexes():
    print(f"  - {index['name']}: {index['key']}")

Indexes on customers:
  - _id_: SON([('_id', 1)])
  - home_zone_1: SON([('home_zone', 1)])
  - complaints.severity_1: SON([('complaints.severity', 1)])
  - complaints.complaint_type_1: SON([('complaints.complaint_type', 1)])

Indexes on deliveries:
  - _id_: SON([('_id', 1)])
  - status_1: SON([('status', 1)])
  - driver_id_1: SON([('driver_id', 1)])
  - hub_id_1: SON([('hub_id', 1)])
  - vehicle_id_1: SON([('vehicle_id', 1)])


Performance After Indexing

In [ ]:
# Same aggregation WITH indexes
start = time.time()
pipeline = [
    {"$unwind": "$complaints"},
    {"$group": {"_id": "$complaints.complaint_type", "count": {"$sum": 1}}},
    {"$sort": {"count": -1}}
]
result = list(customers_col.aggregate(pipeline))
end = time.time()
after_time = round((end - start) * 1000, 2)
print(f"With indexes: {after_time} ms")

print(f"Improvement: {round(before_time - after_time, 2)} ms faster")

With indexes: 40.46 ms
Improvement: 39.02 ms faster


Without indexes: 79.48 ms, With indexes: 40.46 ms. The indexed aggregation runs 39.02 ms faster — a 49% improvement. The query time was nearly halved by adding an index on the field used for grouping. On this dataset of 650 customers the gain is measured in milliseconds, but at production scale with millions of documents, the difference would be seconds.

